SVG patterns might get blurred and pixellated when zooming in

shallawa opened the pull request:
#### 1cc5b31eb1ca1648cd07af6ab48e31db1862ba5e
<pre>
SVG patterns might get blurred and pixellated when zooming in
<a href="https://bugs.webkit.org/show_bug.cgi?id=298753">https://bugs.webkit.org/show_bug.cgi?id=298753</a>
<a href="https://rdar.apple.com/159202567">rdar://159202567</a>

Reviewed by Simon Fraser.

When drawing an SVG pattern, a tileImage is created from the referenced element.
This tileImage is currently scaled by the patternTransform only. So although the
referenced element might be an SVG image or a large bitmap image, the resulted
tileImage will not be scaled well when the coordinate system is scaled like zoom
in or printing for example.

Although the tileImage of the pattern should be scaled by patternTransform, it has
to be scaled by at least GraphicsContext.scaleFactor. This ensures the tileImage
scales properly with the scaled coordinate system.

Test: svg/as-image/svg-as-image-pattern-scale.html
* LayoutTests/svg/as-image/resources/circle-pattern-embeded-image.svg: Added.
* LayoutTests/svg/as-image/svg-as-image-pattern-scale-expected.html: Added.
* LayoutTests/svg/as-image/svg-as-image-pattern-scale.html: Added.
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourcePattern.cpp:
(WebCore::LegacyRenderSVGResourcePattern::createTileImage const):

Canonical link: <a href="https://commits.webkit.org/300357@main">https://commits.webkit.org/300357@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5fe3fa1d3955dba67e262dc57aecc42885f8a985

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/122259 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/41962 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/32631 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/128836 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/74353 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/a9ded4cf-bbda-4920-a2f3-498128a96307) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/124135 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/42676 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/50555 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/92928 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/74353 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/da1ce36a-29ba-4f70-a4e1-905825941378) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/125211 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34040 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/109473 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/73582 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c5713028-5627-499a-b4d1-0f3c01aa7007) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/33049 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/27636 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/72325 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/103548 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/27829 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/131583 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/49198 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/37433 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/101493 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/49573 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/105684 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/101363 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25704 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46731 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/24849 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/45961 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/49055 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/54790 "Failed to build and analyze WebKit") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/48525 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/51875 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/50205 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/7b2e6628b3247af28e98bd67d3235bd611de8acd. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50637

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/78f236b5c2d2f5ddb6b01a726e6f1ea4135ee59c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50637

smfr commented:
> The tileImage of the SVG pattern should be created with the destination GraphicsContext.scaleFactor
> https://bugs.webkit.org/show_bug.cgi?id=298753
> [rdar://159202567](https://rdar.apple.com/159202567)
> 
> Reviewed by NOBODY (OOPS!).
> 
> When drawing an SVG pattern, a tileImage is created from the referenced element.
> This tileImage is currently scaled by the patternTransform only. So although the
> referenced element might be an SVG image or a large bitmap image, the resulted
> tileImage will not scale well when the coordinate system is scaled like zoom in
> or printing for example.
> 
> The tileImage of the pattern should be scaled by the maximum of patternTransform
> and GraphicsContext.scaleFactor. This ensures the tileImage scales properly with
> the scaled coordinate system.

Please describe the user-visibly symptom here somewhere.

shallawa committed:
SVG patterns might get blurred and pixellated when zooming in
https://bugs.webkit.org/show_bug.cgi?id=298753
rdar://159202567

Reviewed by Simon Fraser.

When drawing an SVG pattern, a tileImage is created from the referenced element.
This tileImage is currently scaled by the patternTransform only. So although the
referenced element might be an SVG image or a large bitmap image, the resulted
tileImage will not be scaled well when the coordinate system is scaled like zoom
in or printing for example.

Although the tileImage of the pattern should be scaled by patternTransform, it has
to be scaled by at least GraphicsContext.scaleFactor. This ensures the tileImage
scales properly with the scaled coordinate system.

Test: svg/as-image/svg-as-image-pattern-scale.html
* LayoutTests/svg/as-image/resources/circle-pattern-embeded-image.svg: Added.
* LayoutTests/svg/as-image/svg-as-image-pattern-scale-expected.html: Added.
* LayoutTests/svg/as-image/svg-as-image-pattern-scale.html: Added.
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourcePattern.cpp:
(WebCore::LegacyRenderSVGResourcePattern::createTileImage const):

Canonical link: https://commits.webkit.org/300357@main
